test/e2e/framework.go

```diff
@@ -419,3 +420,136 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
 func KubeDescribe(text string, body func()) bool {
 	return Describe("[k8s.io] "+text, body)
 }
+
+// PodStateVerification represents a verification of pod state.
+// Any time you have a set of pods that you want to operate against or query,
+// this struct can be used to declaratively identify those pods.
+type PodStateVerification struct {
+	// Optional: only pods that have k=v labels will pass this filter.
+	selectors map[string]string
+
+	// Required: The phases which are valid for your pod.
+	validPhases []api.PodPhase
+
+	// Optional: only pods passing this function will pass the filter
+	// Verify a pod.
+	// As an optimization, in addition to specfying filter (boolean),
+	// this function allows specifying an error as well.
+	// The error indicates that the polling of the pod spectrum should stop.
+	verify func(api.Pod) (bool, error)
+
+	// Optional: only pods with this name will pass the filter.
+	podName string
+}
+
+type ClusterVerification struct {
+	framework *Framework
+	podState  PodStateVerification
+}
+
+func (f *Framework) NewClusterVerification(filter PodStateVerification) *ClusterVerification {
+	return &ClusterVerification{
+		f,
+		filter,
+	}
+}
+
+// filter pods passing a filter.  It can be used to compose
+// the more useful abstractions like ForEach, WaitFor, and so on, which
+// can be used directly by tests.
+func (p *PodStateVerification) filter(f *Framework) ([]api.Pod, error) {
+	var pl *api.PodList
+	var err error
+	var selector labels.Selector
+	// List pods based on selectors.  This might be a tiny optimization rather then filtering
+	// everything manually.
+	if len(p.selectors) > 0 {
+		selector = labels.SelectorFromSet(labels.Set(p.selectors))
+		options := api.ListOptions{LabelSelector: selector}
+		pl, err = f.Client.Pods(f.Namespace.Name).List(options)
+	} else {
+		pl, err = f.Client.Pods(f.Namespace.Name).List(api.ListOptions{})
+	}
+	Logf("Selector matched %v pods for %v=%v", len(pl.Items), p.selectors)
+	if len(pl.Items) == 0 || err != nil {
+		return pl.Items, err
+	}
+	// We now have a list of pods that is non empty, carry on with filtering.
+	// We do these filters in separate steps for clarity, extensibility, debugging.
+	pods := pl.Items
+	podFiltered := []api.Pod{}
+	if len(p.validPhases) == 0 {
+		panic("Need to specify a valid pod phase as part of pod states.")
+	}
+	// Next: Pod must match at least one of the states that the user specified
+	for _, pod := range pods {
+		// First Filter: Required
+		podStatePassesFilter := false
+		for _, phase := range p.validPhases {
+			if phase == pod.Status.Phase {
+				podStatePassesFilter = true
+			}
+		}
```

Comment:
we have to check all valid phases so can't have to have the inner for loop ?unless im misunderstanding..